### ISSUE TITLE
How Do I: Use the experience composer to archive a session

### ISSUE THREAD

**Issue by ebarooni** [Vonage/vonage-node-sdk#968]
### How do I

How to use the session, publisher and subscriber event listeners with an experience composer participant

### API/Product

Video

### Code Sample

Hi, i have been trying to use the experience composer for archiving the complete layout of my session. However im a bit stuck with the lack of clear documentation on the topic and unfortunately i was not able to find an example repo that can demonstrate the usage beside this [blog post](https://developer.vonage.com/en/blog/experience-composer-sample-application-with-vonage-video-api).

My sessions heavily use the events dispatched from the `session`, `Publisher` and `Subscriber` objects such as but not limited to`streamCreated`, `streamDestroyed`, `sessionConnected`, `videoElementCreated`. Which then uses the information in the events such the connection id to associate the events to the participants and to organize the session.

My understanding (and challenge) from the experience composer is that once started, it will automatically join the session using the given session id and subscribes to all the participants and the resulting stream from the composer can be used to be published to a new session which then archives the only only available stream (namely the stream of the composer).

But with this approach im not able to setup my listeners for the `Session` (as well as `Subscriber` and `Publisher`) object as i dont i do not construct these objects and the connections myself. Because of this im not able to setup my session layout as it should be on the side of the experience composer.

Is there something im missing from the core concept of using the experience composer or something that i completely understood wrong? I would appreciate if there is a repo with both the front- and backend available where the experience composer is being used to archive the session.

**Comment by ebarooni** [comment]
I forgot to mention that im able to successfully start the experience composer using the REST API. My issue is setting up the layout for the experience composer using the dispatched events.

**Comment by javiermolsanz** [comment]
Hi @ebarooni Javier here from the Solutions engineering team. Let me try and help you out.

Essentially experience composer allows you to publish a new stream to a) either a new session, or b) an existing session. You can then decide depending on your application logic which one is more convenient for you and archive the stream, use it to broadcast or just to inject media into the session like a youtube video in the blog post you mentioned.

Let me know if [this](https://developer.vonage.com/en/blog/why-experience-composer-api-is-great) blog post helps you out. In that blog post, I try to showcase how to publish a custom stream using Experience Composer, but instead of recording that stream in a different session, that stream is published to the same session so that the guest sees a composite view.

Please focus on the [Experience Composer view](https://developer.vonage.com/en/blog/why-experience-composer-api-is-great). In essence, the URL that you pass to the experience composer API has to load code that produces the composite layout that you are looking for. In the blog post scenario I am looking for a layout where the video from the host is present as well as a google calendar and the brand logo.

**Experience Composer Route**
Loads this code which contains the desired layout
https://github.com/nexmo-se/experience-composer-blog/blob/main/src/ec.html
Which loads this JS code
https://github.com/nexmo-se/experience-composer-blog/blob/main/src/index.js

That code does the following:
1. If it's a Experience composer instance, do not publish. You do not want to publish a new stream as experience composer is going to publish whatever finds in the URL you provide, you only need to subscribe. Only publish if this page is [loaded by the host](https://github.com/nexmo-se/experience-composer-blog/blob/main/src/index.js#L110)
2.  If it's a experience composer instance loading this page, subscribe only to the stream published by the host. In you case you will want to publish to every stream as you want a composite containing every stream but this blog post was trying to showcase a different use-case.

**To summarise**
Think of experience composer as a user who joins a website that you provide, executes the logic (JS) found in that website and publish what you would see if you join that website.

That is, for you to publish a composite view of your application you need to have an application route/view where you (do not publish), subscribe to all the streams and display the custom UI elements that you want to see

This is a sample that shows you how to do exactly that https://github.com/nexmo-se/neru-video-experience-composer-archive/tree/main


Let me know if something is not clear, please.

**Comment by ebarooni** [comment]
Hi @javiermolsanz thanks for the clarifications. I was able to finally successfully implement the experience composer in my web application. Nevertheless i appreciate the time you took to clarify things for me. Initially I had several mistakes and misunderstandings in my implementation that lead to my huge confusion. I will try to explain my case below in case it might be helpful for someone else.

My goal was to use the experience composer to archive my session with the complete layout. According to the [REST API reference](https://tokbox.com/developer/rest#starting_experience_composer), to start an experience composer, one must make a post request with the following JSON object included in the request body:

```json
{
  "sessionId": "&lt;session-id&gt;",
  "token": "A valid OpenTok token",
  "url": "https://webapp.customer.com",
  "maxDuration": 1800,
  "resolution": "1280x720",
  "properties": {
    "name": "Composed stream for Live event #1"
  }
}
```

- `sessionId` is explained as "The session ID of the OpenTok session that will include the Experience Composer stream.". Which lead to me thinking i should include the session ID of the session that results from the `url` field. However, I should have used the session ID of a new session, where the experience composer will **publish** its stream to. This new session will also be the session that will be archived.
- As for `token` with a publisher role, I was reusing the token of the moderator of my actual session. Which is completely wrong to do so and was causing conflicts in my own session handling logic. As part of the solution, I generated **two** publisher tokens. One for the request body object, the other one passed as a query parameter to the URL to be able to authenticate the experience composer into my session.
- For the `url`, from reading the documentation and the blogs, my understanding was the experience composer automatically joins the session which is found through the URL (combine this with my explanation on the `sessionId`) and also publishes into the session and subscribes to all the available streams. As a result of this I had disabled all my session handling logic thinking these are done automatically and I was seeing no sign of the experience composer. This also lead to me thinking on how am I supposed to listen for the events emitted from the session and the publisher object for handling my session (hence my question).

The rest of the documentation on the experience composer (stopping, getting info, listing and setting up the callback) is pretty clear and I did not have any issues.

**Comment by javiermolsanz** [comment]
@ebarooni Glad to hear that you managed to implement experience composer in your application and sorry to hear that you found the docs confusing. You feedback is super useful and will definitely pass it on to our team so that we can make the documentation more clear.
